### ISSUE TITLE
okta_group_memberships undoes okta_group_rule users in >=3.26.0

### ISSUE THREAD

**Issue by parente** [okta/terraform-provider-okta#1094]
### Community Note

- Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
- Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
- If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.1.5

### Affected Resource(s)

- okta_group_memberships
- okta_group_rule

### Terraform Configuration Files

```hcl
resource "okta_group_memberships" "group_a_direct" {
  group_id = okta_group.group_a.id
  users = ["00ue640grabcdbA04465", "00u4pjzdcbalzJCP123"] # two users
}

resource "okta_group_rule" "group_b_to_a_rule" {
  name                  = "Group B -> A rule" # say this contains 10 additional users
  status                = "ACTIVE"
  group_assignments     = [okta_group.group_a.id]
  expression_type       = "urn:okta:expression:1.0"
  expression_value      = "isMemberOfAnyGroup(\"${okta_group.group_b.id}\")"
  remove_assigned_users = true
}
```

### Debug Output

N/A

### Panic Output

N/A

### Expected Behavior

Prior to the v3.26.0 release of the okta provider, the two resources depicted above could co-exist without undoing each other in Terraform state.

- The okta_group_rule would ensure all members of group B exist in group A.
- The okta_group_memberships would ensure the two `users` would exist in group A.

### Actual Behavior

Starting in v3.26.0 of the okta provider, the `okta_group_membership` resource imports all members of `okta_group.group_a.id` from Okta into Terraform state for the `okta_group_memberships.group_a_direct` resource on any apply or refresh state action. On the next plan/apply *after* that import, any rule-based users present in the Okta group are deleted from the group by the `okta_group_memberships.group_a_direct` resource since they are present in the state but not present in the `okta_group_memberships` HCL.

### Steps to Reproduce

1. Update provider to 3.26.0 from 3.25.x
2. Plan/apply or refresh state with the HCL declared above.
3. Plan again.
4. Note the plan says all group B users added by `okta_group_rule.group_b_to_a_rule` will be removed by `okta_group_memberships.group_a_direct`.
5. Apply and the plan executes as depicted.

### Important Factoids

The Okta web UI supports groups that have a mix of direct, manual additions as well as rule-based additions. It seems like the same should be true in Terraform.

### References

I think #924 (unintentionally?) introduced this behavior while adding support for `terrform import` of group membership. Now any apply or state refresh updates the okta_group_memberships state with user IDs added by an okta_group_rule.

**Comment by monde** [comment]
Thanks @parente , marking this as regression, I'm trying give these priority.

**Comment by monde** [comment]
@parente I have this PR #1120 that fully documents the `okta_group_memberships`. @exitcode0's #924 actually fixed the resource which in turn exposed questionable behavior your config assumed was true. In your example I'd suggest adding a lifecycle to ignore the `users` argument on `okta_group_memberships`. If anything adds users to Group A outside of the control of `okta_group_memberships.group_a_direct` the resource will appear to drift. There is a ton of explanation demonstrating drift in the acceptance test I wrote for the PR https://github.com/okta/terraform-provider-okta/blob/a0ff118c9ff33abd425136a8befae36605e0e236/okta/resource_okta_group_memberships_test.go#L36-L163

```
resource "okta_group_memberships" "group_a_direct" {
  group_id = okta_group.group_a.id
  users = ["00ue640grabcdbA04465", "00u4pjzdcbalzJCP123"] # two users
  lifecycle {
    ignore_changes = [users]
  }
}

resource "okta_group_rule" "group_b_to_a_rule" {
  name                  = "Group B -> A rule" # say this contains 10 additional users
  status                = "ACTIVE"
  group_assignments     = [okta_group.group_a.id]
  expression_type       = "urn:okta:expression:1.0"
  expression_value      = "isMemberOfAnyGroup(\"${okta_group.group_b.id}\")"
  remove_assigned_users = true
}
```

**Comment by parente** [comment]
@monde Thanks for getting back to me and with a lengthy test case. I agree that `ignore_changes` on `users` will prevent the `okta_group_memberships` resource from removing any users added by the rule. Please correct me if I’m misunderstanding, but won’t that also ignore any future manual changes to the `okta_group_memberships` users list as well? ? If so, would that inhibit our use case of managing users in a group both via direct additions and removals to the group and via rules like we could in the UI?

**Comment by monde** [comment]
@parente imo, okta_group_memberships has to be utilized in one of three ways. First, it is the source of truth, only it adds/removes users, it's state is absolute truth. Second, if users can be added/removed from groups outside of that resource you either need to ignore drift as that is meaningless. Third, import its state periodically (fetching new user ids) before looking at a new plan.

Seems like you want to have it both ways, so the third option is the way for you?

**Comment by parente** [comment]
I definitely want to have it both ways as you say 😁 , but I’m not sure the third way you mention will work. I’d need to leave `users` unignored so that I can continue to add and remove members using `okta_group_memberships`. If an `okta_rule` adds users externally, I can do a `terraform import` to get them in the state. But then the `okta_group_memberships` will immediately remove them on next apply because `users` cannot be ignored. A catch-22 if I’m understanding correctly.

I guess I’m looking for a solution that allows parity with Okta UI clickops prior to 3.26.0, where `okta_group_memberships` and `okta_rule` could both manage members in a single group without trouncing each others’ state. If you say that ship has sailed and the 3.26+ behavior is the new norm, I’ll start looking for ways to work around.

Update: Typo in my original note where I had listed `okta_group_membership` instead of `okta_group_memberships`

**Comment by tmatilai** [comment]
We also want to have a way to add individual users to a group with the `okta_group_memberships` resource, while most users are added with `okta_group_rule` based on their other group memberships. 

If there is no easy/good fix for this, maybe using a collection of `okta_group_membership` (singular) resources would work in this use case? Downside is the required migration of state.

**Comment by tmatilai** [comment]
Oh, the `okta_group_membership` resource has been deprecated. 🙄 

EDIT: which is not stated [in the docs](https://registry.terraform.io/providers/okta/okta/latest/docs/resources/group_membership) btw

**Comment by monde** [comment]
> We also want to have a way to add individual users to a group with the `okta_group_memberships` resource, while most users are added with `okta_group_rule` based on their other group memberships.
> 

@noinarisak @exitcode0 do you all have have the experience if re-importing the user ids for the `okta_group_memberships` if users have been added to the group it represents, but outside its operation of adding said users?

**Comment by monde** [comment]
> Oh, the `okta_group_membership` resource has been deprecated. 🙄
> 
> EDIT: which is not stated [in the docs](https://registry.terraform.io/providers/okta/okta/latest/docs/resources/group_membership) btw

@tmatilai right, when you see a gap in the docs please feel free to do a quick edit and PR. I did so with this one #1122

**Comment by exitcode0** [comment]
> @noinarisak @exitcode0 do you all have have the experience if re-importing the user ids for the `okta_group_memberships` if users have been added to the group it represents, but outside its operation of adding said users?

Not sure i follow what's being asked here. If the resource is `terraform rm`'d and then re-imported the resource in state should have the user IDs for all users currently in the okta group? is this what you were asking @monde?

<hr>

It could be possible to support some additional use-cases here by adding some additional attribute(s) to the `okta_group_memberships` resource similar to `skip_groups`, `skip_users`, and `skip_roles` on other resources.


an attribute like `skip_remove_users` could be added to `okta_group_memberships`.  
if this attribute is set to true, the resource would never remove a user from the group.  
This could facilitate some of this usecase, although im not familiar enough with TF to know if this could be prevented from showing as drift on every plan-apply

It could also make sense to expand the schema for this resource to introduce a concept in state of `users` actually in the Okta group and `managed_users` users for which the resource would take action on

These Ideas could also be implemented in new resources if we didn't want to over-complicate `okta_group_memberships`

**Comment by tmatilai** [comment]
@monde was this issue closed by accident? I don't think the added test offers any resolution to the issue. 😅

**Comment by monde** [comment]
@tmatilai the state and subject of `okta_group_rule` are the properties/expression of the rule itself, not the objects that are affected by the rule. Rules are applied in a job on Okta's backend and are asynchronous, it seems incongruous not to mention error prone to have that resource attempt to keep track of users. The state and subject of `okta_group_memberships` are the users that are in the group directly managed by that resource e.g. `users = [id, id, ...]`. In case of `okta_group_memberships`  I explained and illustrated how drift can occur when users are added to a group outside of the control of the `okta_group_memberships` resource.

@tmatilai is this the bit that you think should keep this issue open? If so, it is its own enhancement request outside of the context of the original issue, it should be in a new issue. That said, as I said, it doesn't make sense to me `okta_group_rule` behave in that manner and it would be error prone.

> We also want to have a way to add individual users to a group with the `okta_group_memberships` resource, while most users are added with `okta_group_rule` based on their other group memberships.
> 
> If there is no easy/good fix for this, maybe using a collection of `okta_group_membership` (singular) resources would work in this use case? Downside is the required migration of state.

**Comment by tmatilai** [comment]
I thought that this issue was about the behaviour change in the `okta_group_memberships` resource in the v3.26.0. That was not a major release, and there was nothing in the changelog, so I would have assumed this indeed was a regression bug as it was initially labeled.

I agree that getting the state though API for the `okta_group_memberships` can be more correct, but the change still broke the existing use case.

For me adding some `skip` attribute to the `okta_group_memberships` resource, as suggested by @exitcode0, would be fine. Or undeprecating the `okta_group_membership` resource. Or any other suitable way.

I guess using one extra group for the individual user memberships, and then adding that to the source group list for the final group would also work, but would lead into extra groups.

**Comment by monde** [comment]
> I thought that this issue was about the behaviour change in the `okta_group_memberships` resource in the v3.26.0. That was not a major release, and there was nothing in the changelog, so I would have assumed this indeed was a regression bug as it was initially labeled.

@tmatilai my mistake, I can see that "fix okta_group_memberships import" #924 was not listed in the v3.26.0 changelog, that was my mistake. As I commented earlier:

> @parente I have this PR #1120 that fully documents the `okta_group_memberships`. @exitcode0's #924 actually fixed the resource which in turn exposed questionable behavior your config assumed was true. 

@tmatilai as this issue thread has gotten long can do me a favor? Can you please open another issue and describe your use case (number of users, groups, apps, etc.). Describe how `okta_group_memberships` behaved previous to v3.26.0 compared to what you see now.  That way we can be focused again. Wee can also discuss bringing back `okta_group_membership`; I did not deprecate it, it was deprecated last summer when `okta_group_memberships` was released. Thank you!

**Comment by monde** [comment]
Update in https://github.com/okta/terraform-provider-okta/releases/tag/v3.29.0 . `okta_group_memberships` reverted back to it's original behavior. If you want it to track all users set the property `track_all_users=true`. Also took extra care to not deviate from the ordering of users ids as they were originally assigned in the update.
https://registry.terraform.io/providers/okta/okta/latest/docs/resources/group_memberships.

Specifically, @parente can you look and see if this delivers your old behavior back?
